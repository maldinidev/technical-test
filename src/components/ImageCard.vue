<template>
  <div class="relative group">
    <img :src="image.download_url" :alt="image.author" @load="handleImageLoad" loading="lazy"
      class="w-full h-48 object-cover shadow transition duration-700 ease-in-out transform"
      :class="image.loaded ? 'opacity-100 scale-100' : 'opacity-0 scale-95'" />
    <div
      class="absolute inset-0 flex items-center justify-center bg-black bg-opacity-50 opacity-0 group-hover:opacity-85 duration-700 ease-in-out cursor-pointer"
      @click="removeImage">
      <span class="text-white font-bold">Eliminar</span>
    </div>
  </div>
</template>

<script setup>
// Define the props for the component
const props = defineProps({
  image: {
    type: Object, // Expecting an object with properties like `download_url`, `author`, and `loaded`
    required: true, // This prop is mandatory for the component to work
  },
});

// Define the events the component can emit
const emit = defineEmits(['remove']); // Declaring the 'remove' event that the parent component will listen to

// Function to mark the image as loaded
const handleImageLoad = () => {
  props.image.loaded = true; // Update the `loaded` property of the image prop when the image finishes loading
};

// Function to emit the 'remove' event to the parent
const removeImage = () => {
  emit('remove', props.image); // Emit the 'remove' event with the image object as the payload
};
</script>
