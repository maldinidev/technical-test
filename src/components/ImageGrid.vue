<template>
  <div>
    <div class="grid grid-cols-2 sm:grid-cols-3 lg:grid-cols-4 gap-4 p-4">
      <ImageCard v-for="image in images" :key="image.id" :image="image" @remove="removeImage" />
    </div>
    <LoadingIndicator :loading="loading" />
  </div>
</template>

<script setup>
import { ref, onMounted, onBeforeUnmount } from 'vue';
import ImageCard from './ImageCard.vue';
import LoadingIndicator from './LoadingIndicator.vue';

// Reactive state variables
const images = ref([]); // Stores the list of images
const loading = ref(false); // Indicates whether data is being loaded
const currentPage = ref(1); // Tracks the current page for pagination
const limit = ref(20); // Number of images to fetch per page

// Fetch images from the API
const fetchImages = async () => {
  try {
    // Set loading state to true
    loading.value = true;
    const response = await fetch(
      `https://picsum.photos/v2/list?page=${currentPage.value}&limit=${limit.value}`
    );
    const data = await response.json();
    // Add new images to the existing list
    images.value.push(...data.map((image) => ({ ...image, loaded: false })));
    // Increment the page for the next fetch
    currentPage.value++;
  } finally {
    // Reset loading state
    loading.value = false;
  }
};

// Remove an image from the list
const removeImage = (image) => {
  const index = images.value.indexOf(image);
  if (index !== -1) {
    // Remove the image from the array
    images.value.splice(index, 1);
  }
};

// Handle infinite scroll
const handleScroll = () => {
  if (
    window.innerHeight + window.scrollY >= document.body.offsetHeight - 100 &&
    !loading.value
  ) {
    fetchImages();
  }
};

// Lifecycle hooks
onMounted(() => {
  // Fetch initial images on component mount
  fetchImages();
  // Add scroll listener
  window.addEventListener('scroll', handleScroll);
});

onBeforeUnmount(() => {
  // Remove scroll listener
  window.removeEventListener('scroll', handleScroll);
});
</script>
