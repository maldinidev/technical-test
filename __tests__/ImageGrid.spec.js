import { mount, flushPromises } from '@vue/test-utils';
import ImageGrid from '../src/components/ImageGrid.vue';
import ImageCard from '../src/components/ImageCard.vue';
import { nextTick } from 'vue';

// Mock the global fetch API
beforeEach(() => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve([
          { id: 1, download_url: 'https://picsum.photos/id/1/200/300', author: 'Author 1', loaded: false },
          { id: 2, download_url: 'https://picsum.photos/id/2/200/300', author: 'Author 2', loaded: false },
        ]),
    })
  );
});

afterEach(() => {
  jest.resetAllMocks();
});

describe('ImageGrid.vue', () => {
  it('renderiza `ImageCard` para cada imagen en el estado', async () => {
    const wrapper = mount(ImageGrid);
    await flushPromises();

    const cards = wrapper.findAllComponents(ImageCard);
    expect(cards.length).toBe(2); // Verifica que hay 2 im치genes en el estado
  });

  it('elimina una imagen del estado cuando se emite "remove"', async () => {
    const wrapper = mount(ImageGrid);

    // Espera a que las im치genes se carguen y el DOM se actualice
    await flushPromises();

    // Verifica el estado inicial
    expect(wrapper.vm.images.length).toBe(2);

    // Simula la emisi칩n del evento "remove" desde el primer `ImageCard`
    const card = wrapper.findComponent(ImageCard);
    const imageToRemove = wrapper.vm.images[0]; // Primera imagen
    await card.vm.$emit('remove', imageToRemove);

    // Espera la actualizaci칩n del estado
    await wrapper.vm.$nextTick();

    // Verifica que el estado se haya actualizado
    expect(wrapper.vm.images.length).toBe(1);
    expect(wrapper.vm.images[0].id).not.toBe(imageToRemove.id);
  });

  it('matches the snapshot', async () => {
    const wrapper = mount(ImageGrid);
    await flushPromises();
    expect(wrapper.html()).toMatchSnapshot();
  });
});
