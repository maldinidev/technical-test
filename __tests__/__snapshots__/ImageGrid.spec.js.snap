// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageGrid.vue matches the snapshot 1`] = `
"<div>
  <div class="grid grid-cols-2 sm:grid-cols-3 lg:grid-cols-4 gap-4 p-4">
    <div class="relative group"><img src="https://picsum.photos/id/1/200/300" alt="Author 1" loading="lazy" class="w-full h-48 object-cover shadow transition duration-700 ease-in-out transform opacity-0 scale-95">
      <div class="absolute inset-0 flex items-center justify-center bg-black bg-opacity-50 opacity-0 group-hover:opacity-85 duration-700 ease-in-out cursor-pointer"><span class="text-white font-bold">Eliminar</span></div>
    </div>
    <div class="relative group"><img src="https://picsum.photos/id/2/200/300" alt="Author 2" loading="lazy" class="w-full h-48 object-cover shadow transition duration-700 ease-in-out transform opacity-0 scale-95">
      <div class="absolute inset-0 flex items-center justify-center bg-black bg-opacity-50 opacity-0 group-hover:opacity-85 duration-700 ease-in-out cursor-pointer"><span class="text-white font-bold">Eliminar</span></div>
    </div>
  </div>
  <!--v-if-->
</div>"
`;
